### PR TITLE
Pull gossip queues for better throughput

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

Lodestar's gossip queues are too restrictive to handle big volumes of objects such when subscribed to all subnets

- See https://github.com/ChainSafe/lodestar/issues/5176

Our existing queues throttle by limiting concurrency, irrespective of downstream worker's capacity. They have the authority to push jobs, and do so independently from each other

**Description**

This PR updates the queue design to:
- Queues just hold data
- NetworkProcessor has the authority to get data from the queues and execute it
- NetworkProcessor throttles based on downstream workers availability to take more work
- Since there's now a central scheduling function it can prioritize topics over others, for example process blocks before attestations

_Marking as draft until testing shows all good_

### CODE REVIEW

**Comment by dapplion** [comment]
Deployed to `feat3`, results show good queue efficiency but mesh health is bad.

_Queues are never full, message drop rate is 0, and job wait time is almost 0_

Screenshot from 2023-02-23 12-15-14

_But mesh peers is now at minimum and don't have peers on most subnets_

Screenshot from 2023-02-23 12-16-36

_We receive way less messages in the beacon_attestation topic_

Screenshot from 2023-02-23 12-17-51

_Peer quality is much different, we have mostly nodes with no validators connected_

Screenshot from 2023-02-23 12-20-10

**EDIT**: From Tuyen

> feat3-lg1k-hzax41 has same issue to beta-lg1k-hzax41, this node contains only low long lived subnets peer, peer count is consistently >=56 peers so there is no chance for new peer. low long lived subnets peer causes low mesh peers see https://github.com/ChainSafe/lodestar/issues/5198


- `feat3-novc-ctvpss`: Queue times for aggregates reduced from 1s to 50ms. Other metrics good.
- `feat3-sm1v-ctvpss`: Queue times for aggregates reduced from 1s to 50ms. Other metrics good.
- `feat3-md16-ctvpsm`: Queue times for aggregates + attestations reduced from 1s to 50ms. Other metrics good.
- `feat3-md64-hzax41`: Affected to issue of peers with low subnet count
- `feat3-lg1k-hzax41`: Affected to issue of peers with low subnet count

**Comment by twoeths** [comment]
peers are not pruned because:
1. most/all peers have duties
2. most/all peers subscribe to >= 1 subnet
3. most/all peers have good score
4. not too grouped to a subnet

this is the same issue to https://github.com/ChainSafe/lodestar/issues/5198

 The reason it happens more to this branch maybe because gossip score it so great (point 3 above - I compared gossipsub score in feat3 and other groups)

File: packages/api/package.json [hunk 1]
```diff
@@ -11,7 +11,7 @@
   "bugs": {
     "url": "https://github.com/ChainSafe/lodestar/issues"
   },
-  "version": "1.4.3",
+  "version": "1.5.0",
```
Comment (twoeths) [inline packages/api/package.json]: I wonder why it shows the diff here while `unstable` really have version as `1.5.0`, should not be an issue anyway

**Comment by twoeths** [comment]
This is awesome work, I can see good metrics like "Job Wait Time" or "Dropped jobs %". However I also notice this push some pressure to I/O lag issue on #4002:

- A little increasement on Event Loop Lag
- A little increasement on Lodestar CPU time (160% -> 170%)
- Especially the `submitPoolAttestations` at validator side (500ms vs 100vs as in stable 1k validator node)
Screen Shot 2023-03-01 at 10 25 47

**Comment by twoeths** [comment]
@dapplion is it possible to introduce some flags to mitigate the issue?

File: packages/api/package.json (hunk 1 above)
Comment (dapplion) [reply packages/api/package.json]: Rebased on unstable and fixed the bad diff

**Comment by dapplion** [comment]
> @dapplion is it possible to introduce some flags to mitigate the issue?

If any this PR is allowing nodes to process all network objects they are required to process as per the spec. "The issue" is actually correct behavior, where current stable drops too many messages. To mitigate "the issue" we should force the node into processing less messages than they can, sort of re-introducing a forced limiting throughput. I'm softly against introducing data, and instead spend our energy understanding what's taking CPU time.

If we were to add flags we could limit the max concurrency per topic.

**Comment by twoeths** [comment]
[0301_gossip-queue-pull_1k.cpuprofile.zip](https://github.com/ChainSafe/lodestar/files/10857742/0301_gossip-queue-pull_1k.cpuprofile.zip)

I'm attaching the profile, there is no special thing except for `handleReceivedMessages()` (then `executeWork`) was called many times inside `runMicroTasks`

In the current implementation we yield to the macro queue every 50ms through `yieldEveryMs` inside `JobItemQueue`, do we have equivalent mechanism in this implementation @dapplion ?

**Comment by dapplion** [comment]
> In the current implementation we yield to the macro queue every 50ms through `yieldEveryMs` inside `JobItemQueue`, do we have equivalent mechanism in this implementation @dapplion ?

That _should_ not be necessary in theory, I added metrics and each gossip packet tends to trigger work on itself. From looking at the profile I don't see issues related with this code specifically. Since everything is run sync now it makes sense to show up on the call stack of handleReceivedMessages and executeWork but their self time is small

File: packages/beacon-node/src/network/processor/worker.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+export type NetworkWorkerModules = ValidatorFnsModules &
+  ValidatorFnModules & {
+    chain: IBeaconChain;
+    events: NetworkEventBus;
+    metrics: Metrics | null;
+    // Optionally pass custom GossipHandlers, for testing
+    gossipHandlers?: GossipHandlers;
+  };
+
+export class NetworkWorker {
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/processor/worker.ts:18]: might be worth adding a tsdoc comment here describing the purpose or mentioning that this isn't related to thread workers

File: packages/beacon-node/src/network/processor/index.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    events: NetworkEventBus;
+    logger: Logger;
+    metrics: Metrics | null;
+  };
+
+export type NetworkProcessorOpts = GossipHandlerOpts & {
+  maxGossipTopicConcurrency?: number;
+};
+
+const executeGossipWorkOrderObj: Record<GossipType, true> = {
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/processor/index.ts:23]: might be worth comment here mentioning the order here defines the order that gossip topics will be processed

File: packages/beacon-node/src/network/processor/index.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+export type NetworkProcessorModules = NetworkWorkerModules &
+  ValidatorFnsModules & {
+    chain: IBeaconChain;
+    events: NetworkEventBus;
+    logger: Logger;
+    metrics: Metrics | null;
+  };
+
+export type NetworkProcessorOpts = GossipHandlerOpts & {
+  maxGossipTopicConcurrency?: number;
```
Comment (wemeetagain) [inline packages/beacon-node/src/network/processor/index.ts:20]: nice to have a tsdoc here

**Review by wemeetagain — APPROVED** [review]
Looks really good. Can you follow up w a PR to add all new metrics to dashboard?
It seems we'll have to tune things based on what we see there.

File: packages/beacon-node/src/network/processor/worker.ts (hunk 2 above)
Comment (dapplion) [reply packages/beacon-node/src/network/processor/worker.ts:18]: Yeah naming sucks, I think this specific class should be merged into the NetworkProcessor

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.6.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.6.0) :tada:

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.7.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.7.0) :tada:

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.8.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.8.0) :tada:

**Pull request merged**